### PR TITLE
Added the ability to combine ranges against the same field under an o…

### PR DESCRIPTION
**Pull request by jwomeara**
…r node into a single ivarator.

This PR also tunes the number of geo ranges that should be generied per tier from 800 to 8.  This value was determined empirically during testing.  

By combining ranges into a single ivarator, and limiting the number of geo ranges per tier to 8, we are able to vastly reduce the runtime for geo queries.

### CODE REVIEW

**Comment by jwomeara** [comment]
Assuming this PR is accepted, I can merge the changes into 2.6 and master.

File: warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexRangeIteratorJexl.java [hunk 1]
```diff
@@ -109,60 +117,90 @@ public String toString() {
...
+                for (Range subRange : subRanges) {
+                    Range range;
+                    if (subRange == subRanges.first()) {
+                        range = new Range(startFi, true, createLowerBoundKey(rowId, fiName, subRange.getStartKey().getRow(), subRange.isStartKeyInclusive()),
+                                        true);
+                        startKey = createUpperBoundKey(rowId, fiName, subRange.getEndKey().getRow(), subRange.isEndKeyInclusive());
+                    } else if (subRange == subRanges.last()) {
+                        range = new Range(startKey, true, endFi, true);
+                    } else {
+                        range = new Range(startKey, true, createLowerBoundKey(rowId, fiName, subRange.getStartKey().getRow(), subRange.isStartKeyInclusive()),
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexRangeIteratorJexl.java]: If we initialized startKey to startFi, I think this if/else block could be simplified to just `if (subRange == subRanges.last()) /* final fi range */ else /* intermediate range */`

Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexRangeIteratorJexl.java]: That's a great suggestion.  Thanks, dude!

File: warehouse/query-core/src/main/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNode.java [hunk 2]
```diff
@@ -45,65 +49,35 @@ public ExceededOrThresholdMarkerJexlNode(JexlNode node) {
...
-        // Create an assignment for the values
-        JexlNode valuesAssignment = JexlNodeFactory.createExpression(JexlNodeFactory.createAssignment(VALUES_PROP,
-                        StringUtils.join(StringUtils.COMMA_STR, new EscapedCollection(values))));
+    public ExceededOrThresholdMarkerJexlNode(String fieldname, Collection<String> values) throws JsonProcessingException {
+        ExceededOrParams params = new ExceededOrParams(fieldname, values);
         
-        // wrap the assignments in an AND node
-        JexlNode andNode = JexlNodeFactory.createUnwrappedAndNode(Arrays.asList(fieldAssignment, valuesAssignment));
+        // Create an assignment for the params
+        JexlNode paramsNode = JexlNodeFactory.createExpression(JexlNodeFactory.createAssignment(EXCEEDED_OR_PARAMS, objectMapper.writeValueAsString(params)));
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNode.java]: All of the constructors here construct an ExceededOrParams and then create the jexlNode and invoke setupSource. We could make one constructor that accepts an ExceededOrParams object, creates the paramsNode, and calls setupSource. Then these constructors could all create the ExceededOrParams and delegate to it. That'd eliminate some duplicated logic.

File: warehouse/query-core/src/main/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNode.java [hunk 3]
```diff
@@ -131,24 +105,113 @@ public static JexlNode getExceededOrThresholdSource(JexlNode node) {
...
+        }
+        
+        private void init(Collection<String> values, Collection<Range> ranges) {
+            if (ranges == null || ranges.isEmpty()) {
+                this.values = values;
+            } else {
+                SortedSet<Range> rangeSet = new TreeSet<>();
+                
+                if (values != null)
+                    values.forEach(value -> rangeSet.add(new Range(new Key(value), new Key(value))));
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNode.java]: `Range.exact(value)` is a more concise way to represent `new Range(new Key(value), new Key(value))` and also avoids the double instantiation.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownLargeFieldedListsVisitor.java [hunk 4]
```diff
@@ -81,96 +95,168 @@ public Object visit(ASTOrNode node, Object data) {
...
+    protected void assignNodeByField(JexlNode origNode, JexlNode subNode, Multimap<String,JexlNode> eqNodes, Multimap<String,JexlNode> rangeNodes,
+                    List<JexlNode> otherNodes) {
+        if (subNode instanceof ASTEQNode) {
+            eqNodes.put(JexlASTHelper.getIdentifier(subNode), origNode);
+        } else if (ExceededValueThresholdMarkerJexlNode.instanceOf(subNode)) {
+            assignNodeByField(origNode, ExceededValueThresholdMarkerJexlNode.getExceededValueThresholdSource(subNode), eqNodes, rangeNodes, otherNodes);
+        }
+        // else if (ASTDelayedPredicate.instanceOf(subNode)) {
+        // assignNodeByField(origNode, ASTDelayedPredicate.getDelayedPredicateSource(subNode), eqNodes, rangeNodes, otherNodes);
+        // }
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownLargeFieldedListsVisitor.java]: remove this

File: warehouse/query-core/src/main/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNode.java [hunk 5]
```diff
@@ -131,24 +105,113 @@ public static JexlNode getExceededOrThresholdSource(JexlNode node) {
...
+                    }
+                }
             }
         }
-        return returnParameters;
+        
+        private Range decodeRange(String[] range) {
+            if (range != null) {
+                if (range.length == 1) {
+                    return new Range(new Key(range[0]), new Key(range[0]));
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNode.java]: This could be `Range.exact(range[0])`

File: warehouse/query-core/src/main/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNode.java [hunk 6]
```diff
@@ -131,24 +105,113 @@ public static JexlNode getExceededOrThresholdSource(JexlNode node) {
...
+                } else if (range.length == 2 && isLowerBoundValid(range[0]) && isUpperBoundValid(range[1])) {
+                    return new Range(new Key(range[0].substring(1)), range[0].charAt(0) == '[', new Key(range[1].substring(0, range[1].length() - 1)),
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNode.java]: For the startKey, I think we need to remove the trailing ']' or ')'.

**Comment by jwomeara** [comment]
Working on adding a test per request from @FineAndDandy

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownLargeFieldedListsVisitor.java (hunk 4 above)
Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownLargeFieldedListsVisitor.java]: Whoops!  Good catch.  Removed in the latest commit.

File: warehouse/query-core/src/main/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNode.java (hunk 5 above)
Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNode.java]: Updated in the latest commit.

File: warehouse/query-core/src/main/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNode.java (hunk 3 above)
Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNode.java]: Nice!  I didn't know about that.  Updated in the latest commit.

File: warehouse/query-core/src/main/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNode.java (hunk 2 above)
Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNode.java]: I kind of wanted to keep the ExceededOrParams class internal to the ExceededOrThresholdMarkerJexlNode to prevent the creation of ExceededOrParams with invalid combinations (i.e. fstPath & values, or fstPath & ranges).  Let me know how you feel about the factory methods I settled on in the latest commit.  I also made the ExceededOrParams class private.

File: warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexRangeIteratorJexl.java [hunk 7]
```diff
@@ -109,60 +117,86 @@ public String toString() {
     
     @Override
     protected List<Range> buildBoundingFiRanges(Text rowId, Text fiName, Text fieldValue) {
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexRangeIteratorJexl.java]: please add unit tests on the boundary conditions of this method

File: warehouse/query-core/src/main/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNode.java (hunk 3 above)
Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNode.java]: Actually, I need to revert this change.  Using Range.exact will add the null byte to the end key, which is something I don't want in this case.  Bummer!

File: warehouse/query-core/src/main/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNode.java (hunk 5 above)
Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNode.java]: Going to revert this change per the comment above.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownLargeFieldedListsVisitor.java [hunk 8]
```diff
@@ -81,96 +95,168 @@ public Object visit(ASTOrNode node, Object data) {
...
+                    }
+                }
+                
+                children.addAll(markers);
             }
             // else simply add the subset back into the children list
             else {
-                children.addAll(subsetChildrenCopies);
+                // recurse on the eq children in this subset
+                for (JexlNode child : eqNodes) {
```
Comment (jschmidt10) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownLargeFieldedListsVisitor.java]: This pattern appears here a few times where we loop over a collection of jexl nodes, apply this visitor, and add the child to the children list if the result is not null. Making a method like `collectChildren(children, eqNodes);` could reduce the duplicated logic in a few places.

File: warehouse/query-core/src/main/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNode.java (hunk 3 above)
Comment (jschmidt10) [reply warehouse/query-core/src/main/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNode.java]: Oh good to know. My bad. Thanks!

File: warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexRangeIteratorJexl.java [hunk 9]
```diff
@@ -109,60 +117,86 @@ public String toString() {
...
+                Key startFi = new Key(rowId, fiName);
+                Key endFi = new Key(rowId, new Text(fiName.toString() + '\0'));
+                
+                Key startKey = startFi;
+                for (Range subRange : subRanges) {
+                    Range range;
+                    if (subRange == subRanges.last()) {
+                        range = new Range(startKey, true, endFi, true);
+                    } else {
+                        range = new Range(startKey, true, createLowerBoundKey(rowId, fiName, subRange.getStartKey().getRow(), subRange.isStartKeyInclusive()),
```
[the remark this answers is not in the document]
Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexRangeIteratorJexl.java]: The ivarator filters on the exact bounded ranges that were passed in, so we should be good in this case.

File: warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexRangeIteratorJexl.java [hunk 10]
```diff
@@ -109,60 +117,86 @@ public String toString() {
...
-            return new RangeSplitter(new Range(startKey, true, endKey, true), getMaxRangeSplit());
-        }
+        return new Key(rowId, fiName, new Text(boundingFiRangeStringBuilder.toString()));
     }
     
     protected Range buildCompositeSafeFiRange(Text rowId, Text fiName, Text fieldValue) {
-        Key startKey = new Key(rowId, fiName, new Text(fieldValue));
-        Key endKey = new Key(rowId, fiName, new Text(upperBound));
-        return new Range(startKey, lowerInclusive, endKey, upperInclusive);
+        if (subRanges != null && !subRanges.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexRangeIteratorJexl.java]: Yeah, it's needed when we do composite seeking.  We need to know the original range that was used so that we can accurately increment the bounds of the ranges as we skip along.

File: warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexRangeIteratorJexl.java (hunk 7 above)
Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/core/iterators/DatawaveFieldIndexRangeIteratorJexl.java]: Got it.  Check out the latest, please.

File: services/base-rest-responses/src/main/java/datawave/webservice/query/exception/DatawaveErrorCode.java [hunk 11]
```diff
@@ -101,7 +101,7 @@
     INDEXED_FIELDS_MISSING(500, 92, "Index lookup found no results or unindexed fields prevented execution of this query."),
     UNEXPECTED_SOURCE_NODE(500, 93, "Unexpected source node."),
     MULTIPLE_RANGES_IN_EXPRESSION(500, 94, "Found multiple ranges in expression."),
-    UNPARSEABLE_URI_PARAMETER(500, 95, "Could not parse URI parameter."),
```
[the remark this answers is not in the document]
Comment (jwomeara) [reply services/base-rest-responses/src/main/java/datawave/webservice/query/exception/DatawaveErrorCode.java]: I don't know of any places where it's used, but I'll sync up with you in person to be sure.

**Comment by jwomeara** [comment]
Rebased the PR on release/version2.6

File: services/base-rest-responses/src/main/java/datawave/webservice/query/exception/DatawaveErrorCode.java [hunk 12]
```diff
@@ -133,7 +133,7 @@
     NODE_EXPANSION_ERROR(500, 124, "Could not expand node."),
     FIELD_NAME_EXPANSION_ERROR(500, 125, "Could not expand field names."),
     METHOD_COMPOSITION_ERROR(500, 126, "Cannot compose method."),
-    FST_CREATE_ERROR(500, 128, "Cannot create FST for large or list"),
```
[the remark this answers is not in the document]
Comment (jwomeara) [reply services/base-rest-responses/src/main/java/datawave/webservice/query/exception/DatawaveErrorCode.java]: I'm sure.

File: services/base-rest-responses/src/main/java/datawave/webservice/query/exception/DatawaveErrorCode.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (jwomeara) [reply services/base-rest-responses/src/main/java/datawave/webservice/query/exception/DatawaveErrorCode.java]: I checked.  It's not used anywhere else that I can find.

File: warehouse/query-core/src/main/java/datawave/query/jexl/DatawaveArithmetic.java [hunk 13]
```diff
@@ -231,6 +236,91 @@ protected Long subtract(Collection<?> left, Collection<?> right) {
...
+                            if (log.isTraceEnabled()) {
+                                log.trace("Failed to evaluate " + o.toString() + " against the FST.");
+                            }
+                        }
+                    }
+                }
+            }
+        } else {
+            if (right != null) {
+                final IntsRefBuilder irBuilder = new IntsRefBuilder();
```
Comment (FineAndDandy) [inline warehouse/query-core/src/main/java/datawave/query/jexl/DatawaveArithmetic.java]: this could be a function to avoid the code copy/paste

File: warehouse/query-core/src/test/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNodeTest.java [hunk 14]
```diff
@@ -0,0 +1,551 @@
...
+import static datawave.webservice.query.QueryParameters.QUERY_AUTHORIZATIONS;
+import static datawave.webservice.query.QueryParameters.QUERY_BEGIN;
+import static datawave.webservice.query.QueryParameters.QUERY_END;
+import static datawave.webservice.query.QueryParameters.QUERY_EXPIRATION;
+import static datawave.webservice.query.QueryParameters.QUERY_NAME;
+import static datawave.webservice.query.QueryParameters.QUERY_PERSISTENCE;
+import static datawave.webservice.query.QueryParameters.QUERY_STRING;
+
+@RunWith(Arquillian.class)
+public class ExceededOrThresholdMarkerJexlNodeTest {
```
Comment (FineAndDandy) [inline warehouse/query-core/src/test/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNodeTest.java]: thanks for adding this

Comment (jwomeara) [reply warehouse/query-core/src/test/java/datawave/query/jexl/nodes/ExceededOrThresholdMarkerJexlNodeTest.java]: Anything for you, FineAndDandy! (y)

File: warehouse/query-core/src/main/java/datawave/query/jexl/DatawaveArithmetic.java (hunk 13 above)
Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/query/jexl/DatawaveArithmetic.java]: I agree.  Updated.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownLargeFieldedListsVisitor.java (hunk 8 above)
Comment (jwomeara) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownLargeFieldedListsVisitor.java]: I agree.  Updated.
